src/cycle.rs

```diff
@@ -1,109 +1,95 @@
-/// user or to help salsa developers figure out why their program is resulting
-/// in a computation cycle.
-///
-/// It is used in a few ways:
-///
-/// * During [cycle recovery](https://https://salsa-rs.github.io/salsa/cycles/fallback.html),
-///   where it is given to the fallback function.
-/// * As the panic value when an unexpected cycle (i.e., a cycle where one or more participants
-///   lacks cycle recovery information) occurs.
-///
-/// You can read more about cycle handling in
-/// the [salsa book](https://https://salsa-rs.github.io/salsa/cycles.html).
-#[derive(Clone, PartialEq, Eq, Hash)]
-pub struct Cycle {
-    participants: CycleParticipants,
+/// Should only be relevant in case of a badly configured cycle recovery.
+pub const MAX_ITERATIONS: u32 = 200;
+
+/// Return value from a cycle recovery function.
+#[derive(Debug)]
+pub enum CycleRecoveryAction<T> {
+    /// Iterate the cycle again to look for a fixpoint.
+    Iterate,
+
+    /// Cut off iteration and use the given result value for this query.
+    Fallback(T),
 }
 
-// We want `Cycle`` to be thin
-pub(crate) type CycleParticipants = Arc<Box<[DatabaseKeyIndex]>>;
+/// Cycle recovery strategy: Is this query capable of recovering from
+/// a cycle that results from executing the function? If so, how?
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum CycleRecoveryStrategy {
+    /// Cannot recover from cycles: panic.
+    ///
+    /// This is the default.
+    Panic,
 
-impl Cycle {
-    pub(crate) fn new(participants: CycleParticipants) -> Self {
-        Self { participants }
-    }
+    /// Recovers from cycles by fixpoint iterating and/or falling
+    /// back to a sentinel value.
+    ///
+    /// This choice is computed by the query's `cycle_recovery`
+    /// function and initial value.
+    Fixpoint,
+}
 
-    /// True if two `Cycle` values represent the same cycle.
-    pub(crate) fn is(&self, cycle: &Cycle) -> bool {
-        Arc::ptr_eq(&self.participants, &cycle.participants)
-    }
+/// A "cycle head" is the query at which we encounter a cycle; that is, if A -> B -> C -> A, then A
+/// would be the cycle head. It returns an "initial value" when the cycle is encountered (if
+/// fixpoint iteration is enabled for that query), and then is responsible for re-iterating the
+/// cycle until it converges. Any provisional value generated by any query in the cycle will track
+/// the cycle head(s) (can be plural in case of nested cycles) representing the cycles it is part
+/// of. This struct tracks these cycle heads.
+#[derive(Clone, Debug, Default)]
+pub(crate) struct CycleHeads(Option<Box<FxHashSet<DatabaseKeyIndex>>>);
 
-    pub(crate) fn throw(self) -> ! {
-        tracing::debug!("throwing cycle {:?}", self);
-        std::panic::resume_unwind(Box::new(self))
+impl CycleHeads {
+    pub(crate) fn is_empty(&self) -> bool {
+        !self.0.as_ref().is_some_and(|heads| !heads.is_empty())
     }
```

Comment:
I think this is a case where we could only test if `self.0` is `Some`. It does require a slightly more complicated `remove` implementation.

Reply:
I can do that but I wouldn't expect it to impact non-cycle-using benchmarks, those should always have `None` already.

Reply:
Done